features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java

```diff
@@ -425,6 +433,23 @@ public Component getComponent() {
         });
     }
 
+    String decodeResourceId(String resourceId){
+        // This is a quite ugly hack but it will be removed as of foundation-2018 when the class ResourceId will be
+        // introduced. We need this because OnmsResource Urlencodes it's id
+        try {
+            return URLDecoder.decode(resourceId, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            // should not happen for UTF-8 but if let's throw an unchecked Exception
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    OnmsResource determineResourceByResourceId(String resourceId){
+        OnmsResource resource = m_resourceDao.getResourceById(resourceId);
+        resource =(resource.getParent()== null) ? resource : resource.getParent();
```

Comment:
Can a resource only have one parent, or do we need to traverse the tree upwards, until no parent is found?

Reply:
Good question about the resources.  I would imagine it's technically possible to have more than one level, but I don't know that we actually do.  Easy enough to fix outside of the context of these changes.